src/content/blog/val-vibes.mdx

```diff
@@ -0,0 +1,166 @@
+
+I started prototyping this on my computer, but then realized: this is Val Town, the ultimate prototyping tool for things like this. There are two phases to semantic search:
+
+1. Indexing: generating embeddings for all public vals.
+2. Querying: generating an embedding for the search query, and comparing to the indexed embeddings.
+
+For indexing, we first create a Postgres table on Neon (using the [pgvector](https://github.com/pgvector/pgvector) extension):
+
+```
+CREATE TABLE vals_embeddings (id TEXT PRIMARY KEY, embedding VECTOR(1536));
+```
+
+Then we create a [cron val](https://docs.val.town/quickstarts/first-cron/) (or as I like to call them: inter-val…), which we set to run every hour.
+
+First we query all vals, for which <a href="https://www.val.town/u/pomdtr">Achille Lacoin</a> had already made an <a href="https://www.val.town/v/sqlite/db">excellent val</a>:
+
+<figure style={{ maxWidth: "600px" }}>
+  <Image src={QueryVals} alt="" />
+  <figcaption>
+    Querying public vals, by <a href="https://www.val.town/u/pomdtr">Achille Lacoin</a>.
+  </figcaption>
+</figure>
+
+We filter out vals for which we have existing embeddings, and make batches of 100 of the remaining vals.
+
+<figure style={{ maxWidth: "700px" }}>
+  <Image src={Indexing1} alt="" />
+  <figcaption>
+  </figcaption>
+</figure>
+
+Then we fetch embeddings from OpenAI for each batch of 100 in parallel, and save them to Postgres.
+
+<figure style={{ maxWidth: "700px" }}>
+  <Image src={Indexing2} alt="" />
+  <figcaption>
+  </figcaption>
+</figure>
+
+This works! Here are the logs from a recent incremental run:
+
+<figure style={{ maxWidth: "700px" }}>
+  <Image src={Indexing3} alt="" />
+  <figcaption>
+  </figcaption>
+</figure>
+
+Querying is simpler still. We turn the query string into an embedding, and sort by similarity:
+
+<figure style={{ maxWidth: "700px" }}>
+  <Image src={Search1} alt="" />
+  <figcaption>
+  </figcaption>
+</figure>
+
+For a nicer UI, I forked an HTTP val (also by <a href="https://www.val.town/u/pomdtr">Achille Lacoin</a>) and hooked it up to my search function:
+
+<figure style={{ maxWidth: "700px" }}>
+  <Image src={Search2} alt="" />
+  <figcaption>
+  </figcaption>
+</figure>
+
+This works pretty well! Check out all my semantic search prototypes [here](https://www.val.town/v/janpaul123/valtownsemanticsearch).
+
+### Product
+
+We have since integrated this into the search of Val Town properly. The implementation is basically the same: we added a pgvector `embedding` column to our Postgres table that stores vals, and we update the embedding whenever changing the code of a val (or its name or readme).
```

Comment:
```suggestion
We have since integrated this into the search of Val Town properly. The implementation is basically the same: we added a pgvector `embedding` column to our Postgres table that stores vals, and we regenerate the embedding whenever a val is updated.
```